### DISCUSSION TITLE
Jpeg compression artifacts in unmasked areas of image2image generation

### DISCUSSION THREAD

**Discussion by Lazy-Bastard** [Comfy-Org/ComfyUI#1779] [General]
I'm doing a lot of iterative image2image generation with masks, and I've noticed that the areas of the image which should stay the same are instead degrading with each generation. It looks like they're undergoing repeated lossy jpeg compression. Is there a way to force ComfyUI to use a lossless image format during this stage? My current workflow has me manually cutting the newly generated image sections into an uncompressed version of the original image with a photo editor and it adds a lot of extra time.

Here's a dark region of an image generated within ComfyUI:
![Original](https://github.com/comfyanonymous/ComfyUI/assets/148297970/831355a8-b99b-407e-8650-3b9d3e3c2745)

And here's that same region after several generations of image2image
![Degraded](https://github.com/comfyanonymous/ComfyUI/assets/148297970/21e09af7-8d5a-4bd7-a1a0-75d5129a9de0)

**Comment by jn-jairo** [comment]
I know what you are talking about, but it is not related to jpeg compression if we are talking about the same thing, it doesn't even use jpeg format, the image used in the nodes is an uncompressed matrix of pixels.

For masked image to image you need to use the ImageCompositeMasked node after to filter the masked area, the KSampler does alter a little bit the parts not masked, for one generation it isn't that noticeable but if you do multiple generations you will see it.

I noticed it on my workflow for upscaled inpaint of masked areas, without the ImageCompositeMasked there is a clear seam on the upscaled square, showing that the whole square image was altered, not just the masked area, but adding the ImageCompositeMasked solved the problem, making a seamless inpaint.

**Reply by ghostsquad** [reply]
@jn-jairo can you provide an example of using ImageCompositeMasked for basic img2img workflow? I'm not sure I understand. Standard img2img only has 2 things really, the latent, and the mask. `ImageCompositeMasked` doesn't even handle latent images. I've been using `Set Latent Noise Mask` node to combine the latent with the mask. I'm curious what I'm missing with all this! 🤔

**Reply by jn-jairo** [reply]
@ghostsquad take a look at this https://github.com/comfyanonymous/ComfyUI/issues/1956#issuecomment-1809334544

**Reply by ghostsquad** [reply]
Hmm, I'm still not sure I'm following along. The previous comment, are you suggesting that after the inpainted image is generated, to grab that section and "paste" it ontop of the original image?

**Reply by ghostsquad** [reply]
Maybe you could share the simplest of example workflows so I can see what you are talking about?

**Reply by jn-jairo** [reply]
@ghostsquad the process of img2img/inpaint always modify the whole image not just the masked area, so to have he same image for the area not maked you need to paste the maked area of the result image on top of the original image, the ImageCompositeMasked do that pasting, other UIs like the Automatic1111 do it automatically in the img2img/inpaint process, to have the same effect in ComfyUI you use the ImageCompositeMasked node.

**Reply by ghostsquad** [reply]
Ah, ok, so there's another node needed to grab the masked area then that needs pasting, such as with this:

https://github.com/BadCafeCode/masquerade-nodes-comfyui#cut-by-mask

I'll see if that solves the problem for me. I'm definitely seeing issues with inpainting and progressive loss of quality.
